File: electron/lib/wsl.js
"use strict";
/* Per-tab WSL runtimes.
 *
 * A session can pin itself to a distro; Studio then keeps one long-lived shell alive
 * for that session so `cd`, env vars and background jobs persist between commands
 * instead of every run starting from scratch. */

const { spawn, execFile } = require("node:child_process");

/** sessionId -> instance */
const instances = new Map();

/** `wsl -l -q` emits UTF-16LE. Decoding it as UTF-8 turns every distro name into
 *  NUL-separated garbage, which then fails every name comparison downstream. */
function decodeWsl(buf) {
  if (buf.length >= 2) {
    let nulls = 0;
    for (let i = 1; i < buf.length; i += 2) if (buf[i] === 0) nulls++;
    if (nulls > buf.length / 4) return buf.toString("utf16le");
  }
  return buf.toString("utf8");
}

function distros() {
  return new Promise((resolve) => {
    execFile(
      "wsl.exe",
      ["-l", "-q"],
      { encoding: "buffer", windowsHide: true, timeout: 15_000 },
      (err, stdout) => {
        if (err && !stdout) {
          resolve([]);
          return;
        }
        resolve(
          decodeWsl(stdout || Buffer.alloc(0))
            .split(/\r?\n/)
            .map((l) => l.replace(/^﻿/, "").trim())
            .filter(Boolean),
        );
      },
    );
  });
}

function toJson(session, inst, status) {
  return {
    session,
    distro: inst.distro,
    cwd: inst.cwd,
    pid: inst.pid,
    startedAt: inst.startedAt,
    auto: inst.auto,
    status,
  };
}

async function list() {
  const out = {};
  for (const [session, inst] of instances) {
    out[session] = toJson(session, inst, inst.exited ? "stopped" : "running");
  }
  return { distros: await distros(), instances: out };
}

async function spawnInstance(args) {
  const available = await distros();
  if (!available.length) throw new Error("no WSL distribution is installed");
  const distro = args.distro || available[0];
  if (!available.includes(distro)) {
    throw new Error(`\`${distro}\` is not an installed WSL distribution`);
  }
  const cwd = args.cwd || "~";
  stopInstance(args.session);
  // `sleep infinity` under a login shell keeps the namespace and the mounted drives
  // alive for the tab without holding a pty open.
  const child = spawn(
    "wsl.exe",
    ["-d", distro, "--cd", cwd, "--", "bash", "-lc", "sleep infinity"],
    { windowsHide: true, stdio: "ignore", detached: false },
  );
  const inst = {
    distro,
    cwd,
    pid: child.pid,
    startedAt: Math.floor(Date.now() / 1000),
    auto: args.auto !== false,
    child,
    exited: false,
  };
  child.on("exit", () => {
    inst.exited = true;
  });
  child.on("error", () => {
    inst.exited = true;
  });
  instances.set(args.session, inst);
  return toJson(args.session, inst, "running");
}

function stopInstance(session) {
  const inst = instances.get(session);
  if (!inst || !inst.child) return false;
  try {
    inst.child.kill();
  } catch {
    /* already gone */
  }
  inst.exited = true;
  return true;
}

function stop(session) {
  const existed = stopInstance(session);
  const inst = instances.get(session);
  return inst
    ? toJson(session, inst, "stopped")
    : { session, status: existed ? "stopped" : "absent" };
}

async function kill(session) {
  stopInstance(session);
  instances.delete(session);
  return list();
}

async function set(session, patch) {
  const inst = instances.get(session);
  if (inst && patch && typeof patch === "object") {
    if (typeof patch.cwd === "string") inst.cwd = patch.cwd;
    if (typeof patch.distro === "string") inst.distro = patch.distro;
    if (typeof patch.auto === "boolean") inst.auto = patch.auto;
  }
  return list();
}

/** Run one command inside the session's distro. Falls back to a one-shot invocation
 *  when no instance is pinned, so a run never silently does nothing. */
async function exec(args) {
  const inst = instances.get(args.session);
  let distro = inst ? inst.distro : args.distro;
  const cwd = inst ? inst.cwd : args.cwd || "~";
  if (!distro) {
    const available = await distros();
    if (!available.length) throw new Error("no WSL distribution is installed");
    distro = available[0];
  }
  const command = args.command || "codex --version";
  return new Promise((resolve) => {
    execFile(
      "wsl.exe",
      ["-d", distro, "--cd", cwd, "--", "bash", "-lc", command],
      { windowsHide: true, maxBuffer: 16 * 1024 * 1024, timeout: 120_000 },
      (err, stdout, stderr) => {
        const lines = [
          { level: "cmd", text: `wsl -d ${distro} --cd ${cwd} -- ${command}` },
        ];
        String(stdout || "")
          .split(/\r?\n/)
          .filter(Boolean)
          .forEach((t) => lines.push({ level: "out", text: t }));
        String(stderr || (err ? err.message : ""))
          .split(/\r?\n/)
          .filter(Boolean)
          .forEach((t) => lines.push({ level: "error", text: t }));
        resolve({
          code: err && typeof err.code === "number" ? err.code : err ? -1 : 0,
          session: args.session,
          distro,
          cwd,
          lines,
        });
      },
    );
  });
}

/** Kill every tracked shell — called when the window closes so a quit does not leave
 *  a `sleep infinity` per tab running forever. */
function shutdown() {
  for (const session of [...instances.keys()]) stopInstance(session);
  instances.clear();
}

module.exports = { distros, list, spawn: spawnInstance, stop, kill, set, exec, shutdown };

File: src-tauri/src/wsl.rs
//! Per-tab WSL runtimes.
//!
//! A session can pin itself to a distro; Studio then keeps one long-lived shell
//! alive for that session so `cd`, env vars and background jobs persist between
//! commands instead of every run starting from scratch.

use crate::cli::command;
use serde::Deserialize;
use serde_json::json;
use serde_json::Value as Json;
use std::collections::HashMap;
use std::process::Child;
use std::process::Stdio;
use std::sync::Mutex;

pub struct Instance {
    pub distro: String,
    pub cwd: String,
    pub pid: u32,
    pub started_at: u64,
    pub auto: bool,
    child: Option<Child>,
}

impl Instance {
    fn to_json(&self, session: &str, status: &str) -> Json {
        json!({
            "session": session,
            "distro": self.distro,
            "cwd": self.cwd,
            "pid": self.pid,
            "startedAt": self.started_at,
            "auto": self.auto,
            "status": status
        })
    }
}

#[derive(Default)]
pub struct Runtimes(pub Mutex<HashMap<String, Instance>>);

#[derive(Deserialize)]
pub struct WslArgs {
    pub session: String,
    pub distro: Option<String>,
    pub cwd: Option<String>,
    pub command: Option<String>,
    pub auto: Option<bool>,
    pub patch: Option<Json>,
}

fn now() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0)
}

/// `wsl -l -q` emits UTF-16LE; decode it rather than lossily reading UTF-8, which
/// would turn every distro name into NUL-separated garbage.
fn decode_wsl(bytes: &[u8]) -> String {
    if bytes.len() >= 2
        && bytes.iter().skip(1).step_by(2).filter(|b| **b == 0).count() > bytes.len() / 4
    {
        let units: Vec<u16> = bytes
            .chunks_exact(2)
            .map(|c| u16::from_le_bytes([c[0], c[1]]))
            .collect();
        return String::from_utf16_lossy(&units);
    }
    String::from_utf8_lossy(bytes).into_owned()
}

pub fn distros() -> Result<Vec<String>, String> {
    let out = command("wsl.exe")
        .args(["-l", "-q"])
        .output()
        .map_err(|e| format!("WSL is not available: {e}"))?;
    Ok(decode_wsl(&out.stdout)
        .lines()
        .map(|l| l.trim().trim_matches('\u{feff}').to_string())
        .filter(|l| !l.is_empty())
        .collect())
}

pub fn list(state: &Runtimes) -> Json {
    let mut map = state.0.lock().unwrap();
    let mut out = serde_json::Map::new();
    for (session, inst) in map.iter_mut() {
        // A `try_wait` that yields Some means the shell exited on its own.
        let status = match inst.child.as_mut().map(|c| c.try_wait()) {
            Some(Ok(Some(_))) | None => "stopped",
            Some(Ok(None)) => "running",
            Some(Err(_)) => "unknown",
        };
        out.insert(session.clone(), inst.to_json(session, status));
    }
    json!({
        "distros": distros().unwrap_or_default(),
        "instances": Json::Object(out)
    })
}

pub fn spawn(state: &Runtimes, args: &WslArgs) -> Result<Json, String> {
    let available = distros()?;
    let distro = args
        .distro
        .clone()
        .or_else(|| available.first().cloned())
        .ok_or("no WSL distribution is installed")?;
    if !available.iter().any(|d| d == &distro) {
        return Err(format!("`{distro}` is not an installed WSL distribution"));
    }
    let cwd = args.cwd.clone().unwrap_or_else(|| "~".to_string());
    // `sleep infinity` under a login shell keeps the namespace and the mounted
    // drives alive for the tab without holding a pty open.
    let child = command("wsl.exe")
        .args([
            "-d",
            &distro,
            "--cd",
            &cwd,
            "--",
            "bash",
            "-lc",
            "sleep infinity",
        ])
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
        .map_err(|e| format!("could not start `{distro}`: {e}"))?;
    let pid = child.id();
    let inst = Instance {
        distro,
        cwd,
        pid,
        started_at: now(),
        auto: args.auto.unwrap_or(true),
        child: Some(child),
    };
    let value = inst.to_json(&args.session, "running");
    stop_inner(state, &args.session);
    state.0.lock().unwrap().insert(args.session.clone(), inst);
    Ok(value)
}

fn stop_inner(state: &Runtimes, session: &str) -> bool {
    let mut map = state.0.lock().unwrap();
    match map.get_mut(session).and_then(|i| i.child.as_mut()) {
        Some(child) => {
            let _ = child.kill();
            let _ = child.wait();
            true
        }
        None => false,
    }
}

pub fn stop(state: &Runtimes, session: &str) -> Json {
    let existed = stop_inner(state, session);
    let map = state.0.lock().unwrap();
    match map.get(session) {
        Some(inst) => inst.to_json(session, "stopped"),
        None => json!({ "session": session, "status": if existed { "stopped" } else { "absent" } }),
    }
}

pub fn kill(state: &Runtimes, session: &str) -> Json {
    stop_inner(state, session);
    state.0.lock().unwrap().remove(session);
    list(state)
}

pub fn set(state: &Runtimes, session: &str, patch: &Json) -> Json {
    {
        let mut map = state.0.lock().unwrap();
        if let Some(inst) = map.get_mut(session) {
            if let Some(cwd) = patch.get("cwd").and_then(|v| v.as_str()) {
                inst.cwd = cwd.to_string();
            }
            if let Some(distro) = patch.get("distro").and_then(|v| v.as_str()) {
                inst.distro = distro.to_string();
            }
            if let Some(auto) = patch.get("auto").and_then(|v| v.as_bool()) {
                inst.auto = auto;
            }
        }
    }
    list(state)
}

/// Run one command inside the session's distro. Falls back to a one-shot
/// invocation when no instance is pinned, so a run never silently does nothing.
pub fn exec(state: &Runtimes, args: &WslArgs) -> Result<Json, String> {
    let (distro, cwd) = {
        let map = state.0.lock().unwrap();
        match map.get(&args.session) {
            Some(inst) => (inst.distro.clone(), inst.cwd.clone()),
            None => (
                args.distro
                    .clone()
                    .or_else(|| distros().ok().and_then(|d| d.first().cloned()))
                    .ok_or("no WSL distribution is installed")?,
                args.cwd.clone().unwrap_or_else(|| "~".to_string()),
            ),
        }
    };
    let cmd_text = args
        .command
        .clone()
        .unwrap_or_else(|| "codex --version".to_string());
    let out = command("wsl.exe")
        .args(["-d", &distro, "--cd", &cwd, "--", "bash", "-lc", &cmd_text])
        .output()
        .map_err(|e| format!("could not run in `{distro}`: {e}"))?;
    let mut lines = vec![
        json!({ "level": "cmd", "text": format!("wsl -d {distro} --cd {cwd} -- {cmd_text}") }),
    ];
    for line in String::from_utf8_lossy(&out.stdout).lines() {
        lines.push(json!({ "level": "out", "text": line }));
    }
    for line in String::from_utf8_lossy(&out.stderr).lines() {
        lines.push(json!({ "level": "error", "text": line }));
    }
    Ok(json!({
        "code": out.status.code().unwrap_or(-1),
        "session": args.session,
        "distro": distro,
        "cwd": cwd,
        "lines": lines
    }))
}
